docs/spec/v2.rst

```diff
@@ -0,0 +1,442 @@
+keys are file names, values are file contents, and files can be read, written 
+or deleted via the operating system. Equally, an S3 bucket can provide this 
+interface, where keys are resource names, values are resource contents, and 
+resources can be read, written or deleted via HTTP.
+
+Below an "array store" refers to any system implementing this interface.
+
+Arrays
+------
+
+Metadata
+~~~~~~~~
+
+Each array requires essential configuration metadata to be stored, enabling 
+correct interpretation of the stored data. This metadata is encoded using JSON 
+and stored as the value of the ".zarray" key within an array store.
+
+The metadata resource is a JSON object. The following keys MUST be present 
+within the object:
+
+zarr_format
+    An integer defining the version of the storage specification to which the
+    array store adheres.
+shape
+    A list of integers defining the length of each dimension of the array.
+chunks
+    A list of integers defining the length of each dimension of a chunk of the
+    array. Note that all chunks within a Zarr array have the same shape.
+dtype
+    A string or list defining a valid data type for the array. See also
+    the subsection below on data type encoding.
+compression
+    A string identifying the primary compression library used to compress
+    each chunk of the array.
+compression_opts
+    An integer, string or dictionary providing options to the primary
+    compression library.
+fill_value
+    A scalar value providing the default value to use for uninitialized
+    portions of the array.
+order
+    Either "C" or "F", defining the layout of bytes within each chunk of the
+    array. "C" means row-major order, i.e., the last dimension varies fastest;
+    "F" means column-major order, i.e., the first dimension varies fastest.
+
+Other keys MUST NOT be present within the metadata object.
+
+For example, the JSON object below defines a 2-dimensional array of 64-bit 
+little-endian floating point numbers with 10000 rows and 10000 columns, divided 
+into chunks of 1000 rows and 1000 columns (so there will be 100 chunks in total 
+arranged in a 10 by 10 grid). Within each chunk the data are laid out in C 
+contiguous order, and each chunk is compressed using the Blosc compression 
+library::
+
+    {
+        "chunks": [
+            1000,
+            1000
+        ],
+        "compression": "blosc",
+        "compression_opts": {
+            "clevel": 5,
+            "cname": "lz4",
+            "shuffle": 1
+        },
+        "dtype": "<f8",
+        "fill_value": null,
+        "order": "C",
+        "shape": [
+            10000,
+            10000
+        ],
+        "zarr_format": 2
+    }
+
+Data type encoding
+~~~~~~~~~~~~~~~~~~
+
+Simple data types are encoded within the array metadata resource as a string, 
+following the `NumPy array protocol type string (typestr) format 
+<http://docs.scipy.org/doc/numpy/reference/arrays.interface.html>`_. The format 
```

Comment:
Good idea, I've expanded the section a bit to include this.

On Friday, 26 August 2016, Stephan Hoyer <notifications@github.com> wrote:

> In docs/spec/v2.rst
> <https://github.com/alimanfoo/zarr/pull/37#discussion_r76455077>:
>
> > +        "dtype": "<f8",
> > +        "fill_value": null,
> > +        "order": "C",
> > +        "shape": [
> > +            10000,
> > +            10000
> > +        ],
> > +        "zarr_format": 2
> > +    }
> > +
> > +Data type encoding
> > +~~~~~~~~~~~~~~~~~~
> > +
> > +Simple data types are encoded within the array metadata resource as a string,
> > +following the `NumPy array protocol type string (typestr) format
> > +<http://docs.scipy.org/doc/numpy/reference/arrays.interface.html>`_. The format
>
> For completeness, I suggest explicitly writing out the list of valid type
> codes and a brief description. I think that would be b, i, u, f, c, S and
> U? I'm not 100% sure it's worth including b or U because both are highly
> inefficiently, though I guess compression should alleviate that somewhat.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alimanfoo/zarr/pull/37/files/8c8dbab434f8cc1fc71d189330d35732564ebf56#r76455077>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAq8QhYdNJVDTaELdMLcRaH4V5Wzlgtyks5qjyDTgaJpZM4JUIhp>
> .
>


-- 
Alistair Miles
Head of Epidemiological Informatics
Centre for Genomics and Global Health <http://cggh.org>
The Wellcome Trust Centre for Human Genetics
Roosevelt Drive
Oxford
OX3 7BN
United Kingdom
Email: alimanfoo@googlemail.com
Web: http://purl.org/net/aliman
Twitter: https://twitter.com/alimanfoo
Tel: +44 (0)1865 287721